tensorflow/python/framework/python_op_gen.cc

```diff
@@ -367,6 +432,108 @@ string GenEagerPythonOp::Code() {
   return prelude_ + result_;
 }
 
+std::unordered_map<string, string> GenEagerPythonOp::GetTypeAnnotationMap() {
+  std::unordered_map<string, string> type_annotations;
+  // Mapping attrs to TypeVars
+  for (const auto& attr : op_def_.attr()) {
+    if (attr.type() == "type") {
+      const string type_var_name = "TV_" + op_def_.name() + "_" + attr.name();
+      type_annotations[attr.name()] = type_var_name;
+    } else if (attr.type() == "bool" || attr.type() == "float" ||
+               attr.type() == "int" || attr.type() == "bytes") {
+      type_annotations[attr.name()] = attr.type();
+    }
+  }
+
+  // Mapping input Tensors to their types
+  for (const auto& arg : op_def_.input_arg()) {
+    // Do not add type annotations to args that accept a sequence of Tensors
+    if (!arg.number_attr().empty()) continue;
+    string type_annotation;
+    if (type_annotations.find(arg.type_attr()) != type_annotations.end()) {
+      // Get the correct TypeVar if input maps to an attr
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", type_annotations[arg.type_attr()], "]");
+    } else {
+      // Get the dtype of the Tensor
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtype_type.find(py_dtype) != dtype_type.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtype_type[py_dtype], "]");
+      }
+    }
+
+    type_annotations[arg.name()] = type_annotation;
+  }
+
+  // Mapping output Tensor to its type
+  if (op_def_.output_arg_size() == 1) {
+    const auto& arg = op_def_.output_arg(0);
+    string type_annotation;
+    if (type_annotations.find(arg.type_attr()) != type_annotations.end()) {
+      strings::StrAppend(&type_annotation, "_ops.Tensor[", type_annotations[arg.type_attr()], "]");
+    } else {
+      const string py_dtype = python_op_gen_internal::DataTypeToPython(arg.type(), "_dtypes.");
+      if (dtype_type.find(py_dtype) != dtype_type.end()) {
+        strings::StrAppend(&type_annotation, "_ops.Tensor[", dtype_type[py_dtype], "]");
+      }
+    }
+
+    type_annotations[arg.name()] = type_annotation;
+  }
+
+  return type_annotations;
+}
+
+// Generate TypeVars using attrs
+void GenEagerPythonOp::GenerateTypeVars(std::unordered_map<string, string>& type_annotations) {
+  bool added_typevar = false;
+  for (const auto& attr : op_def_.attr()) {
+    if (attr.type() == "type") {
+      std::vector<string> allowed_types;
+      for (int t : attr.allowed_values().list().type()) {
+        DataType dtype = static_cast<DataType>(t);
+        const string py_dtype = python_op_gen_internal::DataTypeToPython(dtype, "_dtypes.");
+        if (dtype_type.find(py_dtype) != dtype_type.end()) {
+          allowed_types.emplace_back(dtype_type[py_dtype]);
+        }
+      }
+
+      // If all dtypes are allowed, add them all
```

Comment:
Does this mean "If no dtypes specified, then all are allowed"?

Reply:
Yeah exactly, an example of this is the cast op. The input tensor `x` will have no dtypes specified, so all dtypes all are allowed.

Reply:
SG. Please update the comment.